aiohttp/web_app.py

```diff
@@ -380,15 +383,31 @@ async def _handle(self, request: Request) -> StreamResponse:
             handler = match_info.handler
 
             if self._run_middlewares:
-                for app in match_info.apps[::-1]:
-                    assert app.pre_frozen, "middleware handlers are not ready"
-                    for m in app._middlewares_handlers:
-                        handler = update_wrapper(partial(m, handler=handler), handler)
+                handler = self._apply_middlewares(handler, match_info.apps[::-1])
 
             resp = await handler(request)
 
         return resp
 
+    def _apply_middlewares(
+        self,
+        handler: Handler,
+        apps: Tuple["Application", ...],
+    ) -> Callable[[Request], Awaitable[StreamResponse]]:
+        """Apply middlewares to handler."""
+        cache_key = (handler, tuple(id(app) for app in apps))
```

Comment:
I see no reason we couldn't just define `__hash__` in order to use `@cache`?

Reply:
I was a bit worried about that but as long as we hash on `id` I think its safe

Reply:
I guess, strictly speaking, it's not safe as the ID could be reused for different objects. However, it seems incredibly unlikely to be an issue...